dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,260 @@
+                if (response.code() >= 300) {
+                    OptionalInt next = incrementCurrentHost(currentIndex);
+                    debugLogReceivedErrorStatus(currentIndex, channel, response, next);
+                    // TODO(dfox): handle 308 See Other somehow, as we currently don't have a host -> channel mapping
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                OptionalInt next = incrementCurrentHost(currentIndex);
+                debugLogReceivedThrowable(currentIndex, channel, throwable, next);
+            }
+        }));
+    }
+
+    /**
+     * If we have some reason to think the currentIndex is bad, we want to move to the next host. This is done with a
+     * compareAndSet to ensure that out of order responses which signal information about a previous host don't kick
+     * us off a good one.
+     */
+    private OptionalInt incrementCurrentHost(int currentIndex) {
+        int nextIndex = Math.max(currentIndex + 1, 0); // we want Integer.MAX_VALUE to wrap around to zero
+
+        boolean saved = currentHost.compareAndSet(currentIndex, nextIndex);
+        return saved ? OptionalInt.of(nextIndex) : OptionalInt.empty(); // we've moved on already
+    }
+
+    interface NodeList {
+        /**
+         * Accepts positive indexes that are greater than the length of the list, returns an item modulo the length
+         * of the list.
+         * @return
+         */
+        LimitedChannel get(int index);
+    }
+
+    @VisibleForTesting
+    static final class ConstantNodeList implements NodeList {
+        private final List<LimitedChannel> channels;
+
+        ConstantNodeList(List<LimitedChannel> channels, Random random) {
+            this.channels = shuffleImmutableList(channels, random);
+        }
+
+        @Override
+        public LimitedChannel get(int index) {
+            return channels.get(index % channels.size());
+        }
+    }
+
+    @VisibleForTesting
+    static final class ReshufflingNodeList implements NodeList {
+        private final Ticker clock;
+        private final Random random;
+        private final long intervalWithJitter;
+        private final AtomicLong nextReshuffle;
+
+        private volatile ImmutableList<LimitedChannel> channels;
+
+        ReshufflingNodeList(List<LimitedChannel> channels, Random random, Ticker clock) {
+            this.channels = shuffleImmutableList(channels, random);
+            this.random = random;
+            this.intervalWithJitter = RESHUFFLE_EVERY
+                    .plus(Duration.ofSeconds(random.nextInt(60) - 30))
+                    .toNanos();
+            this.nextReshuffle = new AtomicLong(clock.read() + intervalWithJitter);
+            this.clock = clock;
+        }
+
+        @Override
+        public LimitedChannel get(int index) {
+            reshuffleChannelsIfNecessary();
+            return channels.get(index % channels.size());
```

Comment:
concurrency bug: index out of bounds when another thread reshuffles. We need a single volatile read, and then use get+size on the local reference.

Reply:
How could we get out of bounds if the channel size is always constant? Also we're not shuffling in place, but re-assigning the entire channels list?

Reply:
Ah right, I suppose since we don't have the ability to update targets the size is a constant.

It smells a off that we may be calling channels.size on a different instance than channels.get if a shuffle occurs while we're executing. I think the code would be safer/more obvious with an atomic read (or fully extracting the size) so we don't shoot ourselves in the foot modifying this later!